consumer/src/main/java/com/flipkart/varadhi/consumer/UnGroupedMessageSrc.java

```diff
@@ -0,0 +1,50 @@
+package com.flipkart.varadhi.consumer;
+
+import com.flipkart.varadhi.entities.Offset;
+import com.flipkart.varadhi.spi.services.Consumer;
+import com.flipkart.varadhi.spi.services.PolledMessage;
+import com.flipkart.varadhi.spi.services.PolledMessages;
+import lombok.RequiredArgsConstructor;
+
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.LinkedBlockingQueue;
+
+@RequiredArgsConstructor
+public class UnGroupedMessageSrc<O extends Offset> implements MessageSrc {
+
+    final Consumer<O> consumer;
+
+    private final BlockingQueue<MessageTracker> buffer = new LinkedBlockingQueue<>();
```

Comment:
Does this need to be blocking given the current implementation and caller semantics ?

Reply:
I dont think so. Actually, it probably can be a simple arraylist protected by a lock. synchronziedList

Reply:
single threaded sequential execution is the expectation i.e. multiple nextMessages() concurrent execution is not expected, is synchronization still needed (since buffer is getting manipulated only in nextMessages())

Reply:
`since buffer is getting manipulated only in nextMessages` - not true, when msgSrc does not have any messages, then it will ask the consumer object for more which is async operation, so the new batch of messages will come on a different thread. From that thread, you will get the chance to populate this buffer.

Reply:
Isn't the ArrayList/Linked BlockingQueue roughly the same as list with locks?
Internal implementation is based on Michael and Scott's 2 lock concurrent queue which is designed to be more safe and performant.

Since, at most 2 threads will be accessing it, we can switch to the non-blocking ConcurrentLinkedQueue which is also based on the same paper. (2-lock is more suitable for cases where several threads are contending for shared resource) https://www.cs.rochester.edu/~scott/papers/1996_PODC_queues.pdf

Reply:
PS: I just read this paper few weeks ago, was quite fascinated to find it's mention in the java doc for `ConcurrentLinkedQueue` and `BlockingQueue` implementation. These algorithms are quite ubiquitous!

Reply:
So, these impl are meant to scale array operations when contention due to access from many threads is the issue. Is contention the issue in our case?

I am sure, array with a lock will be more performant in our case.

Actually, I think we dont even need array. Why can't  I just store the iterator reference from the Messages object that I am getting from consumer object?

Reply:
for iterator, we will need a lock when assigning it otherwise it can get overwritten in case due to a race condition.

Reply:
Hmm. I think this can be done without any locks. see if the usage of field `ConcurrencyControlImpl::schedulePendingTaskCounter` inspires you.

Reply:
sure, I'll check that out

Reply:
this is done